src/uwtools/rocoto.py

```diff
@@ -119,7 +119,7 @@ def realize_rocoto_xml(
             # Validate the XML.
             if validate_rocoto_xml(input_xml=temp_file.name):
                 # If no issues were detected, save temp file and report success.
-                shutil.move(temp_file.name, rendered_output)
+                shutil.move(temp_file.name, str(rendered_output))
```

Comment:
I was hoping to replace `shutil.move` with `Path.rename` but since the temp file is a string path, this seemed to work better in this situation

Reply:
reverting this change, so not an issue for this PR

Reply:
This is only going to work this way if the output is a file, no? Wouldn't it need some special treatment if it were stdout?

Reply:
So I thought so, I looked at a few options on file handling, but this seemed to work correctly instantiating the output file. However, if there's a better way I'll happily make it better here!

Reply:
Not quite related to the above discussion, but why use `.name` on `temp_file`, which gives a relative path instead of an absolute one?
```
>>> from pathlib import Path
>>> p = Path("/tmp/foo")
>>> p.is_file()
True
>>> p.name
'foo'
```
Any reason this wouldn't work with just `temp_file` if `shutil.move` accepts a `Path`, or `str(temp_file)` if not?

Or, maybe
```
Path(temp_file).rename(rendered_output)
```
?

Reply:
I'm also curious about the question @christinaholtNOAA raised: Will this work correctly if input is `stdin` and output is `stdout`? The creation of an on-disk temp file would probably best be conditional on the user wanting an on-disk final file; an in-memory / streaming use case should probably not create files at all. It may be functionally correct (and even acceptable) if it does create a temporary file and then remove it, but not as clean as it could be.

Could we see a CLI demo (output pasted here as a comment) of e.g. `cat config.yaml | uw rocoto realize >rocoto.xml` (I'm inventing file names, and maybe other CLI arguments/switches are needed, not sure) and show that `rocoto.xml` is created with the expected content? I'll see if I can try it myself, too.

Reply:
I see what you mean: `temp_file` is not a `Path` object, so `.name` doesn't mean what I thought it did. On the kind of object it **is**, (Python says: `<class 'tempfile._TemporaryFileWrapper'>`), `.name` is indeed the absolute path. Carry on!

Reply:
You're right, of course - it seemed to be ok, but it is creating very wrong after being more rigorous. If I pipe in the input, I get `'NoneType' object has no attribute 'items'`. If I redirect out, it creates a log file called 'None' and gives `ERROR <attribute 'error_log' of 'lxml.etree._Validator' objects>`

Reply:
If I change the input/output definitions in `cli.py` with `type=FileType('w'), default=sys.stdout,` it at least doesn't None me, but I get `[Errno 2] No such file or directory: "<_io.TextIOWrapper name='<stdin>' mode='r' encoding='utf-8'>"`

Reply:
I don't think we'd want to delegate that level of control to `cli.py`. IMO it should just take an optional string, so that `args.output_file` will either the the user-specified path or, if they don't specify one, `None`.

Would a light refactor of `realize_rocoto_xml()` help? I think the code might be clearer anyway without two nested `if` blocks, and what about, just behaviorally speaking:

1. Validate the YAML and return `False` if it fails. (Do we need more logging than what `validate_yaml()` provides? Honest question, I don't know.)
2. Render the template to a temp file.
3. Validate the template and return `False` if it fails.
4. Now that we've got valid XML on disk in the temp file, we can (a) Rename it if argument `rendered_output` is not `None`, or (b) Read it into a string and `print()` it otherwise. We don't have to use `writable()` if it's _less_ convenient.

What do you think?

Reply:
Not really tested, but maybe something like
```
    if not validate_yaml(config_file=config_file, schema_file=_rocoto_schema_yaml()):
        log.error("YAML validation errors identified in %s", config_file)
        return False

    with tempfile.NamedTemporaryFile(delete=False) as temp_file:
        _write_rocoto_xml(config_file=config_file, rendered_output=temp_file.name)

    if not validate_rocoto_xml(input_xml=temp_file.name):
        log.error("Rocoto validation errors identified in %s", temp_file.name)
        return False

    if rendered_output is None:
        with open(temp_file.name, "r", encoding="utf-8") as f:
            print(f.read())
    else:
        Path(temp_file.name).rename(rendered_output)
    return True
```
That depends on import tweaks:
```
-import shutil
+from pathlib import Path
-import uwtools.config.validator
+from uwtools.config.validator import validate_yaml
```
which then wants unit-test patch:
```
% git diff -U0 src/uwtools/tests/test_rocoto.py
diff --git a/src/uwtools/tests/test_rocoto.py b/src/uwtools/tests/test_rocoto.py
index a0031ff..08f1601 100644
--- a/src/uwtools/tests/test_rocoto.py
+++ b/src/uwtools/tests/test_rocoto.py
@@ -13,0 +14 @@ from uwtools import rocoto
+from uwtools.config import validator
@@ -78 +79 @@ def test_realize_rocoto_xml(vals, tmp_path):
-        with patch.object(rocoto.uwtools.config.validator, "_bad_paths", return_value=None):
+        with patch.object(validator, "_bad_paths", return_value=None):
@@ -89 +90 @@ def test_realize_rocoto_invalid_xml():
-        with patch.object(rocoto.uwtools.config.validator, "_bad_paths", return_value=None):
+        with patch.object(validator, "_bad_paths", return_value=None):
```

Reply:
Hrm, it does work with the current tests in this form, but on the command line it still gives me the same errors, unfortunately. I'm not sure we can get away with not handling broader scope arguments in the cli if we want to pipe in and out of it since its expecting `type=str` there

Reply:
> it does work with the current tests in this form, but on the command line it still gives me the same errors
That could mean that the unit tests are not robust enough. Can we devise a unit test that exercises the bug, to avoid regression once we fix it?

As for a fix, could you give me a simple recipe to reproduce the issue via the checked-in code in your fork? I could take a look.

Reply:
I'm pondering how to test piping in the unit tests, I've got a few things to try.

For the errors, in the current code you will need to edit `hello_workflow.yaml` with an explicit path in the `log` tag. For me, line 11 is   `log: /home/vboxuser/Documents/work/workflow-tools/`
I have 3 tests from your suggestion:
1) `cat ~/Documents/work/workflow-tools/src/uwtools/tests/fixtures/hello_workflow.yaml| uw rocoto realize >some.xml`
stdin/stdout - errors with `'NoneType' object has no attribute 'items'`; this seems to be the cli not reading the pipe and setting the object to None.
2) `uw rocoto realize -i ~/Documents/work/workflow-tools/src/uwtools/tests/fixtures/hello_workflow.yaml >~/Documents/work/workflow-tools/src/uwtools/tests/fixtures/some.xml`
stdout only - errors with `ERROR <attribute 'error_log' of 'lxml.etree._Validator' objects>` this seems to be the XML validator failing I think with the inability to read the xml file. I think the redirect is failing so the temp file is creating and passing a null object.
3) `cat ~/Documents/work/workflow-tools/src/uwtools/tests/fixtures/hello_workflow.yaml| uw rocoto realize -o ~/Documents/work/workflow-tools/src/uwtools/tests/fixtures/some.xml`
stdin only - same error as 1

Reply:
Resolved with #325

Reply:
Reopened this conversation as we're having issues in `validate`

Reply:
solved by reverting to `etree.fromstring()`